### ISSUE TITLE
Trumpero - Some rewards will be lost if no floating deposits exist in the market

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-04-interest-rate-model-judging#176]
Trumpero

medium

# Some rewards will be lost if no floating deposits exist in the market

## Summary
When no one deposits into the floating market, the reward could be lost.

## Vulnerability Detail
The `RewardsController` contract is tasked with distributing rewards to users who either deposit or borrow within the protocol. An interesting feature of this contract is the allocation module it employs, which decides how tokens are divided among different user classes (borrowers and depositors) in each period. This is determined by the settings configured for the distribution and is implemented through the `RewardsController.previewAllocation()` function.

```solidity=
  v.depositRewardRule =
    rewardData.depositAllocationWeightAddend.mulWadDown(1e18 - v.sigmoid) +
    rewardData.depositAllocationWeightFactor.mulWadDown(v.sigmoid);
  v.borrowAllocation = v.borrowRewardRule.divWadDown(v.borrowRewardRule + v.depositRewardRule);
  v.depositAllocation = 1e18 - v.borrowAllocation;
  {
    uint256 totalDepositSupply = market.totalSupply();
    uint256 totalBorrowSupply = market.totalFloatingBorrowShares() + m.fixedBorrowShares;
    uint256 baseUnit = distribution[market].baseUnit;
    borrowIndex =
      rewardData.borrowIndex +
      (totalBorrowSupply > 0 ? rewards.mulWadDown(v.borrowAllocation).mulDivDown(baseUnit, totalBorrowSupply) : 0);
    depositIndex =
      rewardData.depositIndex +
      (
        totalDepositSupply > 0
          ? rewards.mulWadDown(v.depositAllocation).mulDivDown(baseUnit, totalDepositSupply)
          : 0
      );
  }
```

After determining the rewards for both the borrowing and depositing sides of the market, the function begins to distribute these rewards by dividing the allocated rewards among the existing shares of the total borrow/deposit supply.

However, issues arise when the `totalBorrowSupply` or `totalDepositSupply` is zero. In such cases, the allocated rewards for that class are effectively lost since there are no shares to receive the rewards. As a result, the index for that class remains unmodified, and no reward accrual takes place for that class, leading to the loss of rewards within the contract.

Upon closer examination, I noticed that when `totalBorrowSupply == 0`, the calculated `target` at [line 500](https://github.com/sherlock-audit/2024-04-interest-rate-model/blob/main/protocol/contracts/RewardsController.sol#L500) also becomes zero, leading to a `distributionFactor` of zero as calculated at [lines 507 - 509](https://github.com/sherlock-audit/2024-04-interest-rate-model/blob/main/protocol/contracts/RewardsController.sol#L507-L509). Consequently, all undistributed rewards from that period are moved to `rewardData.lastUndistributed` and are later redistributed.

However, there is no similar mechanism to address scenarios where `totalDepositSupply == 0` and `totalBorrowSupply != 0`, potentially resulting in reward loss. Such a scenario can occur when users do not engage in the floating market but only participate in depositing and borrowing at specific maturities.

**Proof Of Concept**: 
* Place the test into `protocol/test/M3.t.sol`
* Run the command `forge test --match-test testIssueM3 -vv`

```solidity=
// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.17;

import { Test } from "forge-std/Test.sol";
import { MockERC20 } from "solmate/src/test/utils/mocks/MockERC20.sol";
import { ERC1967Proxy } from "@openzeppelin/contracts/proxy/ERC1967/ERC1967Proxy.sol";
import { FixedPointMathLib } from "solmate/src/utils/FixedPointMathLib.sol";
import { MockInterestRateModel } from "../contracts/mocks/MockInterestRateModel.sol";
import { InterestRateModel } from "../contracts/InterestRateModel.sol";
import { Auditor, IPriceFeed } from "../contracts/Auditor.sol";
import { Market } from "../contracts/Market.sol";
import { MockPriceFeed } from "../contracts/mocks/MockPriceFeed.sol";
import { ERC20, RewardsController, ClaimPermit, InvalidConfig } from "../contracts/RewardsController.sol";
import { FixedLib } from "../contracts/utils/FixedLib.sol";
import "forge-std/console.sol";

contract M3 is Test {
  using FixedPointMathLib for uint256;
  using FixedPointMathLib for int256;

  address internal constant ALICE = address(0x420);
  address internal constant BOB = address(0x69);

  RewardsController internal rewardsController;
  Auditor internal auditor;
  Market internal marketUSDC;
  Market internal marketWETH;
  Market internal marketWBTC;
  MockERC20 internal opRewardAsset;
  MockERC20 internal exaRewardAsset;
  MockInterestRateModel internal irm;

  function setUp() external {
    vm.warp(0);
    MockERC20 usdc = new MockERC20("USD Coin", "USDC", 6);
    MockERC20 weth = new MockERC20("WETH", "WETH", 18);
    MockERC20 wbtc = new MockERC20("WBTC", "WBTC", 8);
    opRewardAsset = new MockERC20("OP", "OP", 18);
    exaRewardAsset = new MockERC20("Exa Reward", "EXA", 18);

    auditor = Auditor(address(new ERC1967Proxy(address(new Auditor(18)), "")));
    auditor.initialize(Auditor.LiquidationIncentive(0.09e18, 0.01e18));
    vm.label(address(auditor), "Auditor");
    irm = new MockInterestRateModel(0.1e18);

    marketUSDC = Market(address(new ERC1967Proxy(address(new Market(usdc, auditor)), "")));
    marketUSDC.initialize(
      "USDC.e",
      3,
      1e18,
      InterestRateModel(address(irm)),
      0.02e18 / uint256(1 days),
      1e17,
      0,
      0.0046e18,
      0.42e18
    );
    vm.label(address(marketUSDC), "MarketUSDC");
    auditor.enableMarket(marketUSDC, new MockPriceFeed(18, 1e18), 0.8e18);

    marketWETH = Market(address(new ERC1967Proxy(address(new Market(weth, auditor)), "")));
    marketWETH.initialize(
      "WETH",
      3,
      1e18,
      InterestRateModel(address(irm)),
      0.02e18 / uint256(1 days),
      1e17,
      0,
      0.0046e18,
      0.42e18
    );
    vm.label(address(marketWETH), "MarketWETH");
    auditor.enableMarket(marketWETH, IPriceFeed(auditor.BASE_FEED()), 0.9e18);

    marketWBTC = Market(address(new ERC1967Proxy(address(new Market(wbtc, auditor)), "")));
    marketWBTC.initialize(
      "WBTC",
      3,
      1e18,
      InterestRateModel(address(irm)),
      0.02e18 / uint256(1 days),
      1e17,
      0,
      0.0046e18,
      0.42e18
    );
    vm.label(address(marketWBTC), "MarketWBTC");
    auditor.enableMarket(marketWBTC, new MockPriceFeed(18, 20_000e18), 0.9e18);

    rewardsController = RewardsController(address(new ERC1967Proxy(address(new RewardsController()), "")));
    rewardsController.initialize();
    vm.label(address(rewardsController), "RewardsController");
    RewardsController.Config[] memory configs = new RewardsController.Config[](1);
    
    configs[0] = RewardsController.Config({
      market: marketWETH,
      reward: opRewardAsset,
      priceFeed: IPriceFeed(address(0)),
      targetDebt: 20_000e6,
      totalDistribution: 2_000 ether,
      start: uint32(block.timestamp),
      distributionPeriod: 12 weeks,
      undistributedFactor: 0.5e18,
      flipSpeed: 2e18,
      compensationFactor: 0.85e18,
      transitionFactor: 0.81e18,
      borrowAllocationWeightFactor: 0,
      depositAllocationWeightAddend: 0.02e18,
      depositAllocationWeightFactor: 0.01e18
    });

    rewardsController.config(configs);
    marketUSDC.setRewardsController(rewardsController);
    marketWETH.setRewardsController(rewardsController);
    opRewardAsset.mint(address(rewardsController), 4_000 ether);
    exaRewardAsset.mint(address(rewardsController), 4_000 ether);

    usdc.mint(address(this), 100 ether);
    usdc.mint(ALICE, 100 ether);
    usdc.mint(BOB, 100 ether);
    weth.mint(address(this), 50_000 ether);
    weth.mint(ALICE, 1_000 ether);
    wbtc.mint(address(this), 1_000e8);
    wbtc.mint(BOB, 1_000e8);
    usdc.approve(address(marketUSDC), type(uint256).max);
    weth.approve(address(marketWETH), type(uint256).max);
    wbtc.approve(address(marketWBTC), type(uint256).max);
    vm.prank(ALICE);
    usdc.approve(address(marketUSDC), type(uint256).max);
    vm.prank(ALICE);
    weth.approve(address(marketWETH), type(uint256).max);
    vm.prank(BOB);
    usdc.approve(address(marketUSDC), type(uint256).max);
    vm.prank(BOB);
    wbtc.approve(address(marketWBTC), type(uint256).max);
  }

  function testIssueM3() external {
    auditor.enterMarket(marketUSDC);

    /// 1. Alice deposits at maturity = FixedLib.INTERVAL
    marketWETH.depositAtMaturity(FixedLib.INTERVAL, 10e18, 10e18, ALICE);

    /// 2. I deposit USDC to marketUSDC as collateral to borrow WETH 
    marketUSDC.deposit(100_000e6, address(this));

    /// 3. I borrow WETH at maturity = FixedLib.INTERVAL 
    ///    use borrowAmount > targetDebt to make sure all reward will be distributed after the distribution conclude
    uint targetDebt = 20_000e6 + 1;
    marketWETH.borrowAtMaturity(FixedLib.INTERVAL, targetDebt, targetDebt * 2, address(this), address(this));

    /// 4. advance time to make the reward distributon conclude 
    (, uint256 distributionEnd, ) = rewardsController.distributionTime(marketWETH, opRewardAsset);
    assertEq(distributionEnd, block.timestamp + 12 weeks);
    vm.warp(distributionEnd + 1000 weeks);

    /// recalculate the actual total distribution (due to truncate)
    uint totalDistribution = 2_000 ether;
    uint distributionPeriod = 12 weeks;
    uint releaseRate = totalDistribution / distributionPeriod;
    uint actualTotalDistribution = releaseRate * distributionPeriod; 
    
    /// 5. I claim rewards 
    (, uint256[] memory claimedAmounts) = rewardsController.claimAll(address(this));
    console.log(claimedAmounts[0]);
    console.log(actualTotalDistribution);
    assertLt(claimedAmounts[0], actualTotalDistribution);
  }
}

```

## Impact
Rewards designated for the deposit class will be lost when `totalDepositSupply` equals zero.

## Code Snippet
https://github.com/sherlock-audit/2024-04-interest-rate-model/blob/main/protocol/contracts/RewardsController.sol#L576-L592

## Tool used
Manual Review

## Recommendation
Consider adding the allocated rewards for the depositors to the `rewardData.lastUndistributed` if the `totalDepositSupply == 0` and `totalBorrowSupply != 0`

**Comment by Dliteofficial** [comment]
Invalid

@santipu03 Why is it important to retain or distribute rewards to users who haven't deposited funds into the floating rate pool? I believe the rewardsController contract is to incentive certain actions such as borrowing and depositing. Also, borrowing cannot happen, if there arent funds to borrow from which means the user has to make some collateral deposit to the pool before they can borrow. At what point is FloatingAsset == 0 that would then necessitate the reward of both the depositors or borrowers?

**Comment by 0xumarkhatab** [comment]
yeah agree with @Dliteofficial  on that .
the finding is confusing , how the total supply can be zero and still the rewards has to be distributed ?

**Comment by santipu03** [comment]
I believe this issue can only be triggered on a market where there are literally 0 floating deposits but some fixed deposits and some fixed borrows. The impact is a loss of some rewards but I agree that the probability of this happening is extremely unlikely because there are always floating deposits on a market.

The only scenario where there can be exactly 0 floating deposits is when a market is deployed. However, the trusted admins won't distribute rewards for a market that has been recently deployed and there are 0 floating deposits there, it makes no sense. 

If someone wants to escalate, this issue may be low severity. 
@Dliteofficial @0xumarkhatab

**Comment by Dliteofficial** [comment]
I'm glad you see why I said this is invalid. I don't think I can escalate because I don't have the threshold to so that. Maybe @0xumarkhatab wants to do that.

**Comment by 0x73696d616f** [comment]
It's not only on deployment is it? If everyone withdraws, it may reach a state where there are no deposits.

**Comment by santipu03** [comment]
> It's not only on deployment is it? If everyone withdraws, it may reach a state where there are no deposits.

Realistically, the deposited assets are strictly zero only on market deployment. Because some users use `withdraw` instead of `redeem`, there's always some dust amount of shares on the market even if all users have abandoned the market. Also, it's less important but there's always the factor that people forget the deposits, lose the seed phrase...

**Comment by Trumpero** [comment]
> The only scenario where there can be exactly 0 floating deposits is when a market is deployed. However, the trusted admins won't distribute rewards for a market that has been recently deployed and there are 0 floating deposits there, it makes no sense. 

I think this scenario still makes sense. The reward distribution is intended to incentivize users to interact with the protocol, so it's likely that the rewards would start when the market is created. I don't understand why the admins wouldn't distribute rewards for a market with zero floating deposits. As I understand it, the rewards are designed for both deposits and borrowing, not just deposits

**Comment by Dliteofficial** [comment]
> > The only scenario where there can be exactly 0 floating deposits is when a market is deployed. However, the trusted admins won't distribute rewards for a market that has been recently deployed and there are 0 floating deposits there, it makes no sense.
> 
> I think this scenario still makes sense. The reward distribution is intended to incentivize users to interact with the protocol, so it's likely that the rewards would start when the market is created. I don't understand why the admins wouldn't distribute rewards for a market with zero floating deposits. As I understand it, the rewards are designed for both deposits and borrowing, not just deposits

The rewards are not there to incentivize interaction rather to reward it. This is where the confusion lies

**Comment by GTH1235** [comment]
Escalate

Check above comments

**Comment by sherlock-admin3** [comment]
> Escalate
> 
> Check above comments

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by Trumpero** [comment]
> The rewards are not there to incentivize interaction rather to reward it. This is where the confusion lies

If you say that the rewards is intended to reward the users who interact with the protocol, so it should distribute for the users who has deposited / borrowed before but in fact the rewards is just starting rewarding since it is created, so it is intended to distribute for the users who are / will interact with the market which in other word incentive ? 

Would love to hear your opinion @santichez

**Comment by santipu03** [comment]
The only possible scenario where this bug can be triggered is when in the market there are some fixed deposits, some fixed borrows, and **strictly 0** floating assets. Realistically, the only moment when the floating assets are strictly zero is at market deployment because later there's always some dust amount even if all users abandon the market. 

**Organic scenario:**
For this bug to be triggered we need some borrows from a fixed pool and deposits from a fixed pool without floating deposits. To borrow, the protocol ensures that there must be liquidity first, so to have fixed borrows we first need to have fixed deposits. The problem here is that no user is going to deposit to a fixed pool if there are still no borrows because he won't earn any interest for it. Therefore, it makes no sense to deposit in a fixed pool when there are no fixed borrows, and there cannot be fixed borrows without some liquidity first. In conclusion, the scenario of 0 floating assets and some fixed borrows is impossible to be given organically.

**Provoked scenario:**
However, an attacker can cause that state when the market is deployed. An attacker could deposit in the floating pool, borrow at maturity, deposit at maturity, and finally redeem all shares from the floating pool. In this scenario, the rewards distributed during the time when there are no floating deposits would be lost. However, this attack makes no sense because the attacker could just deposit some floating assets and start earning all the distributed rewards to that market instead of burning them. In this situation, the attacker could be earning all the distributed rewards because he's the only user in the market, which is the intended design. 

**Admin requirements:**
Also, for this issue to be triggered the admins need to start distributing rewards for a new market when it still doesn't have any floating deposits, which I think makes no sense to do. 

In conclusion, the scenario necessary to trigger this bug cannot be given organically, and it makes no sense for an attacker to cause it because he'd be losing rewards. For these reasons, I think this issue is invalid.

**Comment by santichez** [comment]
Hey @Trumpero 

I agree with @santipu03 's comments. The scenario is really unlikely, also because as a precautionary measure, Exactly team is minting and burning shares at every market deployment to avoid inflation attacks (then floating pool supply will never reach `0`).
The sponsor won't be fixing this finding 👎

**Comment by cvetanovv** [comment]
I agree with the Lead Judge comment. 

Аs he wrote, for this vulnerability to happen, we need fixed deposits, fixed loans, and zero floating assets in the marketplace. But as @santipu03 wrote in the comment, this is almost impossible to happen.

Planning to accept the escalation and invalidate the issue.

**Comment by Evert0x** [comment]
Result:
Invalid
Unique

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [GTH1235](https://github.com/sherlock-audit/2024-04-interest-rate-model-judging/issues/176/#issuecomment-2119654694): accepted
